server/queues/processors/IMIntegrationProcessor.ts

```diff
@@ -0,0 +1,330 @@
+};
+
+export type IntegrationDataProps = {
+  accountIntegration: Integration<IntegrationType.LinkedAccount>;
+  postIntegration: Integration<IntegrationType.Post>;
+  authentication: IntegrationAuthentication;
+};
+
+/** Add the service in IMIntegrationServices (server/utils/integrations.ts) when a new subclass is implemented. */
+export default abstract class IMIntegrationProcessor extends BaseProcessor {
+  static applicableEvents: Event["name"][] = [];
+
+  private service: IntegrationService;
+  private deleteWebhookTaskDelayBound = (5 * Minute) / 1000;
+
+  constructor(service: IntegrationService) {
+    super();
+    this.service = service;
+  }
+
+  public async perform(event: Event): Promise<void> {
+    switch (event.name) {
+      case "documents.publish":
+      case "revisions.create":
+        // wait a few seconds to give the document summary chance to be generated
+        await sleep(5000);
+        return this.documentUpdated(event);
+
+      case "integrations.create":
+        return this.integrationCreated(event);
+
+      case "integrations.delete":
+        return this.integrationDeleted(event);
+
+      default:
+    }
+  }
+
+  private async documentUpdated(
+    event: DocumentEvent | RevisionEvent
+  ): Promise<void> {
+    // don't send notifications when batch importing documents.
+    // @ts-expect-error ts-migrate(2339) FIXME: Property 'data' does not exist on type 'DocumentEv... Remove this comment to see the full error message
+    if (event.data && event.data.source === "import") {
+      return;
+    }
+
+    const [document, team] = await Promise.all([
+      Document.scope(["defaultScope", "withCollection"]).findOne({
+        where: {
+          id: event.documentId,
+        },
+      }),
+      Team.findByPk(event.teamId),
+    ]);
+    if (!document || !team) {
+      return;
+    }
+
+    // don't send notifications for draft documents.
+    if (!document.publishedAt) {
+      return;
+    }
+
+    // if the document was published less than a minute ago,
+    // don't send a separate notification.
+    if (
+      event.name === "revisions.create" &&
+      differenceInMilliseconds(document.updatedAt, document.publishedAt) <
+        Minute
+    ) {
+      return;
+    }
+
+    const integration = (await Integration.findOne({
+      where: {
+        teamId: document.teamId,
+        collectionId: document.collectionId,
+        service: this.service,
+        type: IntegrationType.Post,
+        events: {
+          [Op.contains]: [
+            event.name === "revisions.create" ? "documents.update" : event.name,
+          ],
+        },
+      },
+    })) as Integration<IntegrationType.Post>;
+
+    if (!integration) {
+      return;
+    }
+
+    let text = `${document.updatedBy.name} updated "${document.title}"`;
+
+    if (event.name === "documents.publish") {
+      text = `${document.createdBy.name} published "${document.title}"`;
+    }
+
+    await fetch(integration.settings.url, {
```

Comment:
The abstraction here isn't what I was expecting at all, it seems far too specific to Mattermost. I think you should have abstract methods for posting a message, deleting integration etc and all of the http requests should happen using eg `MattermostClient` in the child implementation.